### PR TITLE
(require|disallow)YodaConditions: add support for listing specific operators

### PR DESCRIPTION
**Pull request by Slayer95**
Fixes https://github.com/jscs-dev/node-jscs/issues/1026

### CODE REVIEW

File: lib/rules/disallow-yoda-conditions.js [hunk 1]
```diff
@@ -1,14 +1,24 @@
...
+ * "disallowYodaConditions": [
+ *     "==",
+ *     "===",
+ *     "!=",
+ *     "!==",
+ *     ">",
+ *     "<",
+ *     ">=",
+ *     "<=",
+ *     "!="
```
Comment (hzoo) [inline lib/rules/disallow-yoda-conditions.js]: there's an extra `!=` here

**Comment by hzoo** [comment]
looks good!

File: lib/rules/disallow-yoda-conditions.js (hunk 1 above)
Comment (Slayer95) [inline lib/rules/disallow-yoda-conditions.js]: Removed it.

**Comment by zxqfox** [comment]
I thought about it. Look like we lost a reason of denying yoda conditions.

Afaik it was a bad code style because there is a difference between `a = 5` and `a == 5`. And mistype is not a syntax error. But it's impossible to mistype `a < 5`. Do we really need this fix or we should just rethink this rule?

**Comment by Slayer95** [comment]
I agree that yoda conditions are only useful to guard against that typo (i.e. <code>requireYodaConditions</code> can be limited to it. However, that's more easily dealt with other sort of rules, like JSHint's <code>"Expected a conditional expression and instead saw an assignment".</code>

Regarding <code>disallowYodaConditions</code>, an argument based on aesthetics and backwards compatibility can be made to keep supporting the current behavior, but I wouldn't mind its getting restricted to <code>==</code>.

Interestingly, ESLint provides the option to forbid Yoda Conditions [except for ranges](http://eslint.org/docs/rules/yoda.html), which is my use case.
<code>requireYodaConditions</code> only needs to distinguish between <code>=</code> and <code>==</code>

File: test/rules/disallow-yoda-conditions.js [hunk 2]
```diff
@@ -3,84 +3,37 @@ var assert = require('assert');
...
+                function() {
+                    checker.configure({ disallowYodaConditions: value });
+                    assert(checker.checkString(notYodaCondition).isEmpty());
+                }
+            );
+        });
     });
 
     it('should report yoda condition for numeric', function() {
+        checker.configure({ disallowYodaConditions: true });
```
Comment (mrjoelkemp) [inline test/rules/disallow-yoda-conditions.js]: Why was this removed from the `beforeEach`?

File: test/rules/require-yoda-conditions.js [hunk 3]
```diff
@@ -3,84 +3,37 @@ var assert = require('assert');
...
+                function() {
+                    checker.configure({ requireYodaConditions: value });
+                    assert(checker.checkString(notYodaCondition).getErrorCount() === 1);
+                }
+            );
+        });
     });
 
     it('should report yoda condition for numeric', function() {
+        checker.configure({ requireYodaConditions: true });
```
Comment (mrjoelkemp) [inline test/rules/require-yoda-conditions.js]: Should go back in the `beforeEach`.

**Comment by mrjoelkemp** [comment]
Is this PR mergeable (once rebased)?

cc @zxqfox @hzoo

**Comment by zxqfox** [comment]
@mrjoelkemp I think it's too complicated and can be simplified. I'm not sure that `5 <= value` should be interpreted as yoda condition.

**Comment by markelog** [comment]
Is that still WIP?

**Comment by Slayer95** [comment]
Waiting for feedback / replies to @zxqof

**Comment by markelog** [comment]
ping @zxqfox

File: lib/rules/disallow-yoda-conditions.js [hunk 4]
```diff
@@ -34,21 +43,22 @@ module.exports = function() {};
...
-            '!=': true,
-            '!==': true,
-            '>': true,
-            '<': true,
-            '>=': true,
-            '<=': true
-        };
+
+        if (isTrue) {
+            operators = ['==', '===', '!=', '!==', '>', '<', '>=', '<='];
```
Comment (zxqfox) [inline lib/rules/disallow-yoda-conditions.js]: What if make ['==', '===', '!='] by default (`true`)?

File: test/specs/rules/require-yoda-conditions.js [hunk 5]
```diff
@@ -3,84 +3,37 @@ var assert = require('assert');
...
+                function() {
+                    checker.configure({ requireYodaConditions: value });
+                    assert(checker.checkString(notYodaCondition).getErrorCount() === 1);
+                }
+            );
+        });
     });
 
     it('should report yoda condition for numeric', function() {
+        checker.configure({ requireYodaConditions: true });
```
Comment (zxqfox) [inline test/specs/rules/require-yoda-conditions.js]: Usually, we making several `describe` blocks with different checker configurations. See https://github.com/jscs-dev/node-jscs/blob/master/test/specs/rules/disallow-space-after-object-keys.js

**Comment by zxqfox** [comment]
Sorry for delay, somehow missed this PR. ;-(

File: lib/rules/disallow-yoda-conditions.js (hunk 4 above)
Comment (mdevils) [inline lib/rules/disallow-yoda-conditions.js]: Why?

Comment (zxqfox) [inline lib/rules/disallow-yoda-conditions.js]: Even just '===' and '=='.
Historical reasons.

Comment (mdevils) [inline lib/rules/disallow-yoda-conditions.js]: I'm not sure about this. As to me they are all Yoda-style. And still I can't see enough reasons to break backwards compatibility.

Comment (Slayer95) [inline lib/rules/disallow-yoda-conditions.js]: It wouldn't actually break anything though. That is, no failing builds for no reason.

File: test/specs/rules/require-yoda-conditions.js (hunk 5 above)
Comment (Slayer95) [inline test/specs/rules/require-yoda-conditions.js]: Will look at it.

File: lib/rules/disallow-yoda-conditions.js (hunk 4 above)
Comment (mdevils) [inline lib/rules/disallow-yoda-conditions.js]: I'm not talking about the builds. I'm about the unexpected behaviour change. For instance, I know several projects which rely on the current behaviour.

Comment (zxqfox) [inline lib/rules/disallow-yoda-conditions.js]: Actually I'm fine if we left it as is.

**Comment by Slayer95** [comment]
Updated tests as requested.

**Comment by zxqfox** [comment]
@Slayer95 Thanks!

LGTM

**Comment by mdevils** [comment]
LGTM! Good job!

**Comment by markelog** [comment]
Just a remainder, if pr has two "LGTM" you can merge it, @mdevils, @zxqfox could one of you do that?

**Comment by mdevils** [comment]
@markelog Yes, sure :)

**Comment by mdevils** [comment]
Merged into `2.0`. Thank you!
